### ISSUE TITLE
Security Issues with Nginx (1.21.1-alpine)

### ISSUE THREAD

**Issue by shahiinn** [nginxinc/docker-nginx#564]
On scanning the latest version of Nginx (1.21.1-alpine) against JFrog Xray we could see there are multiple security issues.
````
   "total_count": 12,
   "data": [
      {
         "id": "",
         "severity": "Critical",
         "severity_source": "CVSS V3 from RBS",
         "summary": "Libxslt transform.c xsltApplyTemplates() Function Node Deletion Use-after-free Arbitrary Code Execution",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:libxslt",
         "source_id": "alpine://3.14:libxslt",
         "source_comp_id": "alpine://3.14:libxslt:1.1.34-r1",
         "component_versions": {
            "id": "3.14:libxslt",
            "vulnerable_versions": [
               "≤ 1.1.34-r1"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "9.3/CVSS:2.0/AV:N/AC:M/Au:N/C:C/I:C/A:C",
                     "cvss_v3": "9.8/CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H"
                  },
                  {
                     "cve": "CVE-2021-30560"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-28T09:27:10Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Critical",
         "severity_source": "CVSS V3 from RBS",
         "summary": "FreeType psaux/psintrp.c cf2_interpT2CharString() Function Integer Underflow Unspecified Issue",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:freetype",
         "source_id": "alpine://3.14:freetype",
         "source_comp_id": "alpine://3.14:freetype:2.10.4-r1",
         "component_versions": {
            "id": "3.14:freetype",
            "vulnerable_versions": [
               "≤ 2.10.4-r1"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "9.3/CVSS:2.0/AV:N/AC:M/Au:N/C:C/I:C/A:C",
                     "cvss_v3": "9.8/CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-23T09:27:13Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "High",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl Metalink Feature Download Handling Remote Credential Disclosure Weakness",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:libcurl",
         "source_id": "alpine://3.14:libcurl",
         "source_comp_id": "alpine://3.14:libcurl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:libcurl",
            "vulnerable_versions": [
               "7.46.0-r2 ≤ Version < 7.78.0-r0"
            ],
            "fixed_versions": [
               "7.78.0-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "4.3/CVSS:2.0/AV:N/AC:M/Au:N/C:P/I:N/A:N",
                     "cvss_v3": "8.6/CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:C/C:H/I:N/A:N"
                  },
                  {
                     "cve": "CVE-2021-22923"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-28T09:27:10Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "High",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl Metalink Feature Download Handling Remote Credential Disclosure Weakness",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:curl",
         "source_id": "alpine://3.14:curl",
         "source_comp_id": "alpine://3.14:curl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:curl",
            "vulnerable_versions": [
               "7.27.0-r0 ≤ Version < 7.78.0-r0"
            ],
            "fixed_versions": [
               "7.78.0-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "4.3/CVSS:2.0/AV:N/AC:M/Au:N/C:P/I:N/A:N",
                     "cvss_v3": "8.6/CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:C/C:H/I:N/A:N"
                  },
                  {
                     "cve": "CVE-2021-22923"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-28T09:27:10Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "High",
         "severity_source": "CVSS V3 from RBS",
         "summary": "Google WebP (libwebp) mux/anim_encode.c WebPAnimEncoderAdd() Function GIF Image Handling Out-of-bounds Read Issue",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:libwebp",
         "source_id": "alpine://3.14:libwebp",
         "source_comp_id": "alpine://3.14:libwebp:1.2.0-r2",
         "component_versions": {
            "id": "3.14:libwebp",
            "vulnerable_versions": [
               "≤ 1.2.0-r2"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "7.8/CVSS:2.0/AV:N/AC:M/Au:N/C:P/I:N/A:C",
                     "cvss_v3": "8.2/CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:L/I:N/A:H"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-02T09:26:14Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "High",
         "severity_source": "CVSS V3 from RBS",
         "summary": "Google WebP (libwebp) utils/rescaler_utils.c WebPRescalerInit() Function Integer Overflow Buffer Overflow DoS",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:libwebp",
         "source_id": "alpine://3.14:libwebp",
         "source_comp_id": "alpine://3.14:libwebp:1.2.0-r2",
         "component_versions": {
            "id": "3.14:libwebp",
            "vulnerable_versions": [
               "≤ 1.2.0-r2"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "7.1/CVSS:2.0/AV:N/AC:M/Au:N/C:N/I:N/A:C",
                     "cvss_v3": "7.5/CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-02T09:26:14Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "High",
         "severity_source": "CVSS V3 from RBS",
         "summary": "FreeType psaux/psintrp.c cf2_interpT2CharString() Function Integer Underflow Out-of-bounds Access DoS",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:freetype",
         "source_id": "alpine://3.14:freetype",
         "source_comp_id": "alpine://3.14:freetype:2.10.4-r1",
         "component_versions": {
            "id": "3.14:freetype",
            "vulnerable_versions": [
               "≤ 2.10.4-r1"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "7.1/CVSS:2.0/AV:N/AC:M/Au:N/C:N/I:N/A:C",
                     "cvss_v3": "7.5/CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-19T09:27:14Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "High",
         "severity_source": "CVSS V3 from RBS",
         "summary": "FreeType psnames/psmodule.c ps_unicodes_init() Function Empty Glyph Name Handling Out-of-bounds Read Issue",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:freetype",
         "source_id": "alpine://3.14:freetype",
         "source_comp_id": "alpine://3.14:freetype:2.10.4-r1",
         "component_versions": {
            "id": "3.14:freetype",
            "vulnerable_versions": [
               "≤ 2.10.4-r1"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "7.8/AV:N/AC:M/Au:N/C:P/I:N/A:C",
                     "cvss_v3": "8.2/CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:L/I:N/A:H"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-19T09:27:14Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Medium",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl lib/vtls/vtls.c Curl_ssl_config_matches() Function Improper Path Name Check Connection Reuse Weakness",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:libcurl",
         "source_id": "alpine://3.14:libcurl",
         "source_comp_id": "alpine://3.14:libcurl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:libcurl",
            "vulnerable_versions": [
               "7.46.0-r2 ≤ Version < 7.78.0-r0"
            ],
            "fixed_versions": [
               "7.78.0-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:N/I:P/A:N",
                     "cvss_v3": "6.5/CVSS:3.0/AV:N/AC:H/PR:N/UI:N/S:U/C:H/I:L/A:N"
                  },
                  {
                     "cve": "CVE-2021-22924"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-28T09:27:10Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Medium",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl lib/vtls/vtls.c Curl_ssl_config_matches() Function Improper Path Name Check Connection Reuse Weakness",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:curl",
         "source_id": "alpine://3.14:curl",
         "source_comp_id": "alpine://3.14:curl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:curl",
            "vulnerable_versions": [
               "7.19.2-r0 ≤ Version < 7.78.0-r0"
            ],
            "fixed_versions": [
               "7.78.0-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:N/I:P/A:N",
                     "cvss_v3": "6.5/CVSS:3.0/AV:N/AC:H/PR:N/UI:N/S:U/C:H/I:L/A:N"
                  },
                  {
                     "cve": "CVE-2021-22924"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-28T09:27:10Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Medium",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl Metalink Feature Content Download Improper Hash Mismatch Handling Weakness",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:curl",
         "source_id": "alpine://3.14:curl",
         "source_comp_id": "alpine://3.14:curl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:curl",
            "vulnerable_versions": [
               "7.27.0-r0 ≤ Version < 7.78.0-r0"
            ],
            "fixed_versions": [
               "7.78.0-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:N/I:P/A:N",
                     "cvss_v3": "5.9/CVSS:3.0/AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:H/A:N"
                  },
                  {
                     "cve": "CVE-2021-22922"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-28T09:27:10Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Medium",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl Metalink Feature Content Download Improper Hash Mismatch Handling Weakness",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:libcurl",
         "source_id": "alpine://3.14:libcurl",
         "source_comp_id": "alpine://3.14:libcurl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:libcurl",
            "vulnerable_versions": [
               "7.46.0-r2 ≤ Version < 7.78.0-r0"
            ],
            "fixed_versions": [
               "7.78.0-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:N/I:P/A:N",
                     "cvss_v3": "5.9/CVSS:3.0/AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:H/A:N"
                  },
                  {
                     "cve": "CVE-2021-22922"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-28T09:27:10Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Low",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl lib/telnet.c suboption() Function telnet Handshake Uninitialized Memory Disclosure",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:curl",
         "source_id": "alpine://3.14:curl",
         "source_comp_id": "alpine://3.14:curl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:curl",
            "vulnerable_versions": [
               "7.19.2-r0 ≤ Version < 7.78.0-r0"
            ],
            "fixed_versions": [
               "7.78.0-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:P/I:N/A:N",
                     "cvss_v3": "3.7/CVSS:3.0/AV:N/AC:H/PR:N/UI:N/S:U/C:L/I:N/A:N"
                  },
                  {
                     "cve": "CVE-2021-22898",
                     "cwe": [
                        "CWE-909"
                     ],
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:P/I:N/A:N",
                     "cvss_v3": "3.1/CVSS:3.1/AV:N/AC:H/PR:N/UI:R/S:U/C:L/I:N/A:N"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-27T09:27:12Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Low",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl lib/telnet.c suboption() Function telnet Handshake Uninitialized Memory Disclosure",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:libcurl",
         "source_id": "alpine://3.14:libcurl",
         "source_comp_id": "alpine://3.14:libcurl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:libcurl",
            "vulnerable_versions": [
               "7.46.0-r2 ≤ Version < 7.78.0-r0"
            ],
            "fixed_versions": [
               "7.78.0-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:P/I:N/A:N",
                     "cvss_v3": "3.7/CVSS:3.0/AV:N/AC:H/PR:N/UI:N/S:U/C:L/I:N/A:N"
                  },
                  {
                     "cve": "CVE-2021-22898",
                     "cwe": [
                        "CWE-909"
                     ],
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:P/I:N/A:N",
                     "cvss_v3": "3.1/CVSS:3.1/AV:N/AC:H/PR:N/UI:R/S:U/C:L/I:N/A:N"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-27T09:27:12Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Low",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl lib/vtls/sectransp.c sectransp_connect_step1() Function Client Certificate Confusion Local Issue",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:curl",
         "source_id": "alpine://3.14:curl",
         "source_comp_id": "alpine://3.14:curl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:curl",
            "vulnerable_versions": [
               "7.33.0-r0 ≤ Version < 7.78.0-r0"
            ],
            "fixed_versions": [
               "7.78.0-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:N/I:P/A:N",
                     "cvss_v3": "3.7/CVSS:3.0/AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:L/A:N"
                  },
                  {
                     "cve": "CVE-2021-22926"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-27T09:27:12Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Low",
         "severity_source": "CVSS V3 from RBS",
         "summary": "cURL / libcurl lib/vtls/sectransp.c sectransp_connect_step1() Function Client Certificate Confusion Local Issue",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:libcurl",
         "source_id": "alpine://3.14:libcurl",
         "source_comp_id": "alpine://3.14:libcurl:7.77.0-r1",
         "component_versions": {
            "id": "3.14:libcurl",
            "vulnerable_versions": [
               "7.46.0-r2 ≤ Version ≤ 7.77.0-r1"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "2.6/CVSS:2.0/AV:N/AC:H/Au:N/C:N/I:P/A:N",
                     "cvss_v3": "3.7/CVSS:3.0/AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:L/A:N"
                  },
                  {
                     "cve": "CVE-2021-22926"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-27T09:27:12Z",
         "is_source_root": false
      },
      {
         "id": "",
         "severity": "Unknown",
         "severity_source": "CVSS V3 from RBS",
         "summary": "Perl-Compatible Regular Expressions (PCRE) Regular Expression Handling Memory Consumption DoS Weakness",
         "issue_type": "security",
         "provider": "JFrog",
         "component": "3.14:pcre",
         "source_id": "alpine://3.14:pcre",
         "source_comp_id": "alpine://3.14:pcre:8.44-r0",
         "component_versions": {
            "id": "3.14:pcre",
            "vulnerable_versions": [
               "8.43-r0 ≤ Version ≤ 8.44-r0"
            ],
            "more_details": {
               "cves": [
                  {
                     "cvss_v2": "0.0/AV:N/AC:M/Au:N/C:N/I:N/A:N",
                     "cvss_v3": "0.0/CVSS:3.0/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:N"
                  }
               ],
               "provider": "JFrog"
            }
         },
         "edited": "2021-07-19T09:27:01Z",
         "is_source_root": false
      }
   ]
}```

Can you please have a look into the issues and provide an update.

**Comment by shahiinn** [comment]
raised it under the [wrong project](https://github.com/nginxinc/docker-nginx-unprivileged/issues/73) earlier and hence opening a new one here. I have scanned the base image of alpine for vulnerabilities and could see there is just 1 medium issue and the rest of the Vulnerabilities are part of the Nginx image alone.
Curl has a new version available for alpine and so all the issues with curl or libcurl should be fixed in the next release I guess.

**Comment by shahiinn** [comment]
From the [Dockerfile](https://github.com/nginxinc/docker-nginx/blob/master/mainline/alpine/Dockerfile#L50) it seems below are the parent packages which fetches the Vulnerable coponents.

pcre-dev - pcre
gd-dev - freetype
gd-dev - libwebp
libxslt-dev - libxslt

Can these packages be removed or replaced?

**Comment by yosifkit** [comment]
Background: 
>Tags in the [official-images] library file[s] are only built through an update to that library file or as a result of its base image being updated (ie, an image `FROM debian:buster` would be rebuilt when `debian:buster` is built). 
>
>\-https://github.com/docker-library/official-images/tree/2f086314307c04e1de77f0a515f20671e60d40bb#library-definition-files

Official Images FAQ:

> Though not every CVE is removed from the images, we take CVEs seriously and try to ensure that images contain the most up-to-date packages available within a reasonable time frame
>
> \- https://github.com/docker-library/faq/tree/6138d05aabf61563606d86f98d0ccbd99f162b33#why-does-my-security-scanner-show-that-an-image-has-cves

Since our build system makes heavy use of Docker build cache, just rebuilding the all of the alpine-based `nginx` Dockerfiles won't cause any change. So we rely on periodic base image updates.

> We strive to publish updated images at least monthly for Debian and Ubuntu. We also rebuild earlier if there is a critical security need. Many Official Images are maintained by the community or their respective upstream projects, like Alpine and Oracle Linux, and are subject to their own maintenance schedule.
>
> \- from the same FAQ link

----

Fortunately, the `alpine:3.14` images were just updated in https://github.com/docker-library/official-images/pull/10674 and so rebuilds are in progress. So if there are any fixes that can be pulled in via `apk`, they'll be part of the newly built images.

**Comment by thresheek** [comment]
Unfortunately it looks like none of the affected packages, except, notably, `curl`, were updated to fix those issues in Alpine 3.14.

**Comment by thresheek** [comment]
Checked `nginx:1.21.3-alpine` and it's the same still - no updates from Alpine to the affected packages (libxslt, freetype, curl, libwebp, pcre).

**Comment by Moulick** [comment]
any timeline when these vulnerabilities will be closed ?

**Comment by thresheek** [comment]
Hi @Moulick, no idea - we don't control Alpine Linux security fixes.

**Comment by thresheek** [comment]
JFYI, this is still the same for `1.21.4` tags, no packages have been fixed security-wise in Alpine Linux, except curl.

**Comment by thresheek** [comment]
This is now fixed in `nginx:1.21.5-alpine`, which uses `alpine:3.15`.

We now ship:
- libxslt-1.1.34-r1
- freetype-2.11.0-r0
- curl-7.80.0-r0
- libwebp-1.2.1-r0
- pcre2-10.39-r0

which have all those issues fixed.
